## Load Enriched Data & Calculate Baseline Metrics

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("../data/processed/pipeline_output.csv")

print(f"Total reviews loaded: {len(df)}")
print("\n--- Average Rating per Bank ---")
print(df.groupby('bank')['rating'].mean())

print("\n--- Sentiment Distribution per Bank ---")
print(df.groupby(['bank', 'sentiment_label']).size())

print("\n--- Top Identified Themes per Bank ---")
print(df.groupby(['bank', 'identified_theme']).size())

Total reviews loaded: 968

--- Average Rating per Bank ---
bank
Bank of Abyssinia              3.353659
Commercial Bank of Ethiopia    3.684887
Dashen Bank                    3.905775
Name: rating, dtype: float64

--- Sentiment Distribution per Bank ---
bank                         sentiment_label
Bank of Abyssinia            negative           181
                             positive           147
Commercial Bank of Ethiopia  negative           144
                             positive           167
Dashen Bank                  negative           137
                             positive           192
dtype: int64

--- Top Identified Themes per Bank ---
bank                         identified_theme                 
Bank of Abyssinia            Account Access Issues                 24
                             Customer Support & Bank Specifics     16
                             General Feedback                     147
                             Transaction Performance           



---

### Key Baseline Insights & Findings

* **Overall Ratings & Market Standing**:
* **Dashen Bank** leads the group with the highest average rating of **3.91**, indicating relatively better user satisfaction and app reliability compared to its peers.
* **Commercial Bank of Ethiopia (CBE)** sits in the middle with a **3.68** average rating, reflecting a mix of strong transaction adoption alongside recurring performance feedback.
* **Bank of Abyssinia (BOA)** trails behind with a **3.35** average rating, driven by a higher volume of negative sentiment relative to positive reviews (181 negative vs. 147 positive).


* **Sentiment vs. Rating Dynamics**:
* BOA's negative sentiment skew directly correlates with its lower star rating average.
* Conversely, Dashen Bank and CBE enjoy higher positive review counts (192 and 167 positive reviews respectively), showing stronger baseline feature appreciation despite common fintech pain points.

---

## Extracting Satisfaction Drivers and Pain Points per Bank

In [27]:
# Cell: Extract drivers (positive) and pain points (negative) per bank with theme evidence
for bank_name in df['bank'].unique():
    print(f"                   BANK: {bank_name}")

    
    bank_df = df[df['bank'] == bank_name]
    
    print("\n          Top Positive Themes (Drivers):    ")
    pos_themes = bank_df[bank_df['sentiment_label'] == 'positive']['identified_theme'].value_counts()
    print(pos_themes.head(3))
    
    print("\n             Top Negative Themes (Pain Points):")
    neg_themes = bank_df[bank_df['sentiment_label'] == 'negative']['identified_theme'].value_counts()
    print(neg_themes.head(3))
    print(" ")
    

    sample_pos = bank_df[(bank_df['bank'] == bank_name) & (bank_df['sentiment_label'] == 'positive')]['review_text'].values
    if len(sample_pos) > 0:
        print(f'Sample Positive Review (Driver Evidence): "{sample_pos[1]}"')
        
    sample_neg = bank_df[(bank_df['bank'] == bank_name) & (bank_df['sentiment_label'] == 'negative')]['review_text'].values
    if len(sample_neg) > 0:
        print(f'Sample Negative Review (Pain Point Evidence): "{sample_neg[1]}"')
    print("--------------------------------------------------------------------------\n")

                   BANK: Commercial Bank of Ethiopia

          Top Positive Themes (Drivers):    
identified_theme
General Feedback                     79
UI & Design & App Stability          67
Customer Support & Bank Specifics    11
Name: count, dtype: int64

             Top Negative Themes (Pain Points):
identified_theme
General Feedback               56
UI & Design & App Stability    46
Account Access Issues          26
Name: count, dtype: int64
 
Sample Positive Review (Driver Evidence): "ok. easly updare"
Sample Negative Review (Pain Point Evidence): "The app stops while log in"
--------------------------------------------------------------------------

                   BANK: Bank of Abyssinia

          Top Positive Themes (Drivers):    
identified_theme
General Feedback                     76
UI & Design & App Stability          51
Customer Support & Bank Specifics    12
Name: count, dtype: int64

             Top Negative Themes (Pain Points):
identified_theme
General Feed


---

### Satisfaction Drivers & Pain Points Analysis (Evidence-Backed)

#### 1. Commercial Bank of Ethiopia (CBE)

* **Satisfaction Drivers**:
* *Evidence*: Users frequently praise app updates and general usability under the `UI & Design & App Stability` theme (67 positive mentions).
* *Sample Evidence*: `"ok. easly updare"`


* **Pain Points**:
* *Evidence*: A high concentration of negative feedback targets app stability and access hurdles (46 mentions for UI/Stability and 26 for Account Access Issues).
* *Sample Evidence*: `"The app stops while log in"` (highlighting critical session/authentication dropouts).



#### 2. Bank of Abyssinia (BOA)

* **Satisfaction Drivers**:
* *Evidence*: Strong appreciation for core general banking utilities (`General Feedback` has 76 positive instances, alongside 51 positive stability mentions).
* *Sample Evidence*: `"thanks You"`


* **Pain Points**:
* *Evidence*: Transaction performance and stability are major friction points, showing high negative counts under `UI & Design & App Stability` (54 mentions) and `Transaction Performance` (33 mentions).
* *Sample Evidence*: While brief, negative context aligns with lagging transaction processing and operational errors.



#### 3. Dashen Bank

* **Satisfaction Drivers**:
* *Evidence*: Leads in positive user sentiment regarding interface design and functional execution (75 positive mentions for `UI & Design & App Stability` and 91 for general satisfaction).
* *Sample Evidence*: `"nice done"`


* **Pain Points**:
* *Evidence*: Notable negative volume arises from account accessibility and inclusive design gaps (25 mentions under `Account Access Issues`).
* *Sample Evidence*: `"accessibility for people with visual impairment should be given more consideration"` (highlighting a specific UI/UX inclusivity barrier).

---

## Compare banks across sentiment distribution, average ratings, and dominant themes

In [28]:
print("1. AVERAGE RATINGS PER BANK")
print(df.groupby('bank')['rating'].mean().round(2))
print(" ")

print("2. SENTIMENT DISTRIBUTION (%) PER BANK")
sentiment_dist = pd.crosstab(df['bank'], df['sentiment_label'], normalize='index') * 100
print(sentiment_dist.round(2))
print(" ")

print("3. DOMINANT THEMES PER BANK")
dominant_themes = pd.crosstab(df['bank'], df['identified_theme'])
print(dominant_themes)

1. AVERAGE RATINGS PER BANK
bank
Bank of Abyssinia              3.35
Commercial Bank of Ethiopia    3.68
Dashen Bank                    3.91
Name: rating, dtype: float64
 
2. SENTIMENT DISTRIBUTION (%) PER BANK
sentiment_label              negative  positive
bank                                           
Bank of Abyssinia               55.18     44.82
Commercial Bank of Ethiopia     46.30     53.70
Dashen Bank                     41.64     58.36
 
3. DOMINANT THEMES PER BANK
identified_theme             Account Access Issues  \
bank                                                 
Bank of Abyssinia                               24   
Commercial Bank of Ethiopia                     30   
Dashen Bank                                     32   

identified_theme             Customer Support & Bank Specifics  \
bank                                                             
Bank of Abyssinia                                           16   
Commercial Bank of Ethiopia                       


---

### Comparative Analysis Across Key Dimensions

#### 1. Average Rating & Market Standing

* **Dashen Bank** leads the competitive group with an average user rating of **3.91 / 5.00**, aligning with its superior positive sentiment balance.
* **Commercial Bank of Ethiopia (CBE)** occupies the mid-tier position with an average rating of **3.68 / 5.00**, reflecting steady transaction usage offset by mixed stability performance.
* **Bank of Abyssinia (BOA)** trails its peers with an average rating of **3.35 / 5.00**, indicating deeper underlying friction points impacting user perception.

#### 2. Sentiment Distribution (%)

* **Dashen Bank**: Demonstrates the healthiest sentiment ratio, with **58.36% positive** reviews versus **41.64% negative** reviews.
* **Commercial Bank of Ethiopia**: Maintains a positive majority at **53.70%**, though closely balanced by a **46.30% negative** share.
* **Bank of Abyssinia**: Experiences a negative skew, where **55.18%** of reviews express negative sentiment compared to **44.82%** positive.

#### 3. Dominant Themes & Operational Impact

* **General Feedback**: Forms the largest volume block across all institutions (~135–147 mentions per bank), capturing broad app utilization comments.
* **UI & Design & App Stability**: Represents the second most prominent operational cluster (ranging from 105 to 113 mentions per bank), serving as a primary driver for both praise and crash reports.
* **Account Access Issues & Transaction Performance**: Serve as critical technical pain points, with Bank of Abyssinia registering the highest transactional friction volume (36 mentions) and notable account access constraints across all three platforms.

---




# Data-Grounded Product & Support Improvements

#### 1. Commercial Bank of Ethiopia (CBE)

* **Improvement 1: Authentication & Session Stability Patch**
* *Data Justification*: Addressed frequent crash and freeze reports during login sequences (noted under `UI & Design & App Stability` and access dropouts).
* *Action*: Optimize background authentication token handling and session management to prevent abrupt app termination during launch.


* **Improvement 2: Seamless In-App Update Architecture**
* *Data Justification*: Positive feedback centers heavily on smooth updates (`"ok. easly updare"`), but failure points disrupt experience.
* *Action*: Implement silent background patching and clearer progress indicators to maintain high satisfaction during version upgrades.



#### 2. Bank of Abyssinia (BOA)

* **Improvement 1: Transaction Processing & Queue Optimization**
* *Data Justification*: BOA suffers from a high concentration of negative feedback regarding transaction performance and system lags (`Transaction Performance` pain points).
* *Action*: Scale server-side API capacity and implement clear asynchronous status states for pending or slow transfers to reduce user frustration.


* **Improvement 2: Proactive Customer Support Integration**
* *Data Justification*: Lower overall satisfaction ratings (3.35 average) and high negative review skew (55.18%) indicate a need for direct support channels.
* *Action*: Introduce an in-app live chat or automated ticketing helper to resolve account and transaction errors before customers resort to negative public store reviews.



#### 3. Dashen Bank

* **Improvement 1: Inclusive Design & Accessibility Overhaul**
* *Data Justification*: Direct qualitative evidence highlights accessibility barriers ("accessibility for people with visual impairment should be given more consideration").
* *Action*: Implement WCAG-compliant UI enhancements, including adjustable font scaling, high-contrast modes, and full screen-reader compatibility.


* **Improvement 2: Streamlined Account Recovery & Onboarding**
* *Data Justification*: Despite leading with a 3.91 rating, account access issues remain a recurring technical complaint category.
* *Action*: Simplify self-service PIN resets and biometric re-authentication flows to minimize friction for returning users.
---